OpenRA.Mods.Cnc/Traits/Conditions/GrantConditionByDisguisedActor.cs

```diff
@@ -0,0 +1,68 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Mods.Common.Traits.Render;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Cnc.Traits.Render
+{
+	class GrantConditionByDisguisedActorInfo : ITraitInfo, Requires<DisguiseInfo>
+	{
+		[Desc("Conditions to grant when disguised as specified actor.",
+			"A dictionary of [actor id]: [condition].")]
+		public readonly Dictionary<string, string> Conditions = new Dictionary<string, string>();
```

Comment:
I think it would make more sense to have this defined on the `Disguise` trait itself as a `DisguiseConditions` dictionary.

Reply:
HitShape's `LoadShape` is a good example of what I mean, but if you can get the per-actor stuff working on a branch then i'm happy to show you the specific code changes to make the generic condition work.

Reply:
In fact, i tried to add that into `Disguise:`first, but couldn't get it work. Here is the old code, if you can see what is wrong with it, i can use it for this pr: https://github.com/MustaphaTR/OpenRA/commit/743a05a095898ba1f3ab05f28e5974dae6e43af4

IIRC, with this code revoking the condition didn't work. So i was getting multipile conditions after disguising as several units.

Reply:
You added your `disguisedAsToken` handling code inside an `if (Disguised != oldDisguiseSetting)`, which means it only runs when going from not disguised &lrarr; disguised, and not when changing the disguise type.